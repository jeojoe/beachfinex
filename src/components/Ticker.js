import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { actionCreators } from '../reducers/ticker';

const Pair = styled.h1`
  color: #fff;
  margin-left: 5px;
  margin-right: 30px;
`;

const Label = styled.p`
  color: #fff;
`;

const Value = styled.p`
  color: #fff;
`;

class Ticker extends React.Component {
  constructor(props) {
    super(props);
    this.ws = new WebSocket('wss://api.bitfinex.com/ws/2');
    this.ws.onopen = this.onOpen;
    this.ws.onmessage = this.onMessage;
    this.ws.onerror = this.onError;
  }

  onOpen = () => {
    const openMsg = JSON.stringify({
      event: 'subscribe',
      channel: 'ticker',
      symbol: 'tBTCUSD',
    });
    this.ws.send(openMsg);
  }

  onMessage = (msg) => {
    const { setTicker } = this.props;
    const data = JSON.parse(msg.data)[1];
    if (data && data !== 'hb') {
      setTicker(data);
    }
  }

  onError = err => console.log(err);

  render() {
    const { ticker } = this.props;

    if (!ticker.lastPrice) return 'Fetching..';

    return (
      <div className="row">
        <Pair>BTC/USD</Pair>
        <div className="col-1">
          <Label>Last Price</Label>
          <Value>{ticker.lastPrice}</Value>
        </div>
        <div className="col-1">
          <Label>Change 24hr</Label>
          <Value>{ticker.dailyChange} ({ticker.dailyChangePerc})</Value>
        </div>
        <div className="col-1">
          <Label>High 24hr</Label>
          <Value>{ticker.high}</Value>
        </div>
        <div className="col-1">
          <Label>Low 24hr</Label>
          <Value>{ticker.low}</Value>
        </div>
        <div className="col-1">
          <Label>Volume 24hr</Label>
          <Value>{ticker.volume}</Value>
        </div>
      </div>
    );
  }
}

Ticker.propTypes = {
  ticker: PropTypes.shape({
    bid: PropTypes.number,
    bidSize: PropTypes.number,
    ask: PropTypes.number,
    askSize: PropTypes.number,
    dailyChange: PropTypes.number,
    dailyChangePerc: PropTypes.number,
    lastPrice: PropTypes.number,
    volume: PropTypes.number,
    high: PropTypes.number,
    low: PropTypes.number,
  }).isRequired,
  setTicker: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    ticker: state.ticker,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setTicker: rawTicker => dispatch(actionCreators.setTicker(rawTicker)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Ticker);
