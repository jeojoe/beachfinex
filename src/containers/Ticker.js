import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { actionCreators } from '../reducers/ticker';
import BTC from '../images/BTC.png';

const Logo = styled.img`
  width: 50px;
  height: 50px;
  margin-top: 15px;
  margin-left: 20px;
`;

const Pair = styled.h1`
  color: #fff;
  margin-left: 15px;
  background: url()
`;

const Label = styled.p`
  color: #fff;
  font-size: 12px;
  font-weight: bold;
  margin-bottom: 10px;
`;

const Value = styled.p`
  color: ${props => props.color || '#fff'};
  display: inline;
  margin-top: 0;
  font-size: 20px;
  font-weight: bold;
  white-space: nowrap;
`;

export class Ticker extends React.Component {
  componentDidMount() {
    document.title = 'BTC-USD | Beachfinex';
    this.ws = new WebSocket('wss://api.bitfinex.com/ws/2');
    this.ws.onopen = this.onOpen;
    this.ws.onmessage = this.onMessage;
    this.ws.onerror = this.onError;
  }

  onOpen = () => {
    const openMsg = JSON.stringify({
      event: 'subscribe',
      channel: 'ticker',
      symbol: 'tBTCUSD',
    });
    this.ws.send(openMsg);
  }

  onMessage = (msg) => {
    const { setTicker } = this.props;
    const data = JSON.parse(msg.data)[1];
    if (data && data !== 'hb') {
      setTicker(data);
    }
  }

  onError = err => console.log(err);

  render() {
    const {
      ticker: {
        dailyChange,
        dailyChangePerc,
        lastPrice,
        volume,
        high,
        low,
      },
    } = this.props;

    if (!lastPrice) return 'Fetching..';

    document.title = `BTC-USD ${lastPrice.toFixed(1)} | Beachfinex`;

    return (
      <div className="row justify-between">
        <div className="row">
          <Logo src={BTC} alt="BTC Logo" width="50" height="50" />
          <Pair>BTC/USD</Pair>
        </div>
        <div className="row align-center">
          <div className="col-1 ticker">
            <Label>Last Price</Label>
            <Value>{lastPrice.toFixed(1)}</Value>
          </div>
          <div className="col-1 ticker">
            <Label>Change 24hr</Label>
            <Value
              color={dailyChange >= 0 ? 'lime' : 'red'}
            >
              {dailyChange.toFixed(2)} ({dailyChangePerc.toFixed(2)}%)
            </Value>
          </div>
          <div className="col-1 ticker">
            <Label>High 24hr</Label>
            <Value>{high.toFixed(1)}</Value>
          </div>
          <div className="col-1 ticker">
            <Label>Low 24hr</Label>
            <Value>{low.toFixed(1)}</Value>
          </div>
          <div className="col-1 ticker">
            <Label>Volume 24hr</Label>
            <Value>{volume.toFixed(2)}</Value>
          </div>
        </div>
      </div>
    );
  }
}

Ticker.propTypes = {
  ticker: PropTypes.shape({
    bid: PropTypes.number,
    bidSize: PropTypes.number,
    ask: PropTypes.number,
    askSize: PropTypes.number,
    dailyChange: PropTypes.number,
    dailyChangePerc: PropTypes.number,
    lastPrice: PropTypes.number,
    volume: PropTypes.number,
    high: PropTypes.number,
    low: PropTypes.number,
  }).isRequired,
  setTicker: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    ticker: state.ticker,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setTicker: rawTicker => dispatch(actionCreators.setTicker(rawTicker)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Ticker);
