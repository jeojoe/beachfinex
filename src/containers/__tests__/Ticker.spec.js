import React from 'react';
import { shallow } from 'enzyme';

import { Ticker } from '../Ticker';

describe('<Ticker />', () => {
  it('renders correctly', () => {
    const setTicker = jest.fn();
    const ticker = {
      bid: 0,
      bidSize: 1,
      ask: 2,
      askSize: 3,
      dailyChange: 4,
      dailyChangePerc: 5,
      lastPrice: 6,
      volume: 7,
      high: 8,
      low: 9,
    };
    const wrapped = shallow(
      <Ticker
        ticker={ticker}
        setTicker={setTicker}
        ws={{
          subscribed: false,
          subscribing: true,
          subscribeSuccess: jest.fn(),
          subscribe: jest.fn(),
          unsubscribe: jest.fn(),
          toggle: jest.fn(),
          setMoreActions: jest.fn(),
        }}
      />,
    );
    expect(wrapped).toMatchSnapshot();
  });
});
