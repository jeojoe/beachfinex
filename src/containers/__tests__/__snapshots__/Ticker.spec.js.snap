// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ticker /> renders correctly 1`] = `
<div
  className="row justify-between"
>
  <div
    className="row"
  >
    <styled.img
      alt="BTC Logo"
      height="50"
      src="BTC.png"
      width="50"
    />
    <styled.h1>
      BTC/USD
    </styled.h1>
  </div>
  <div
    className="row align-center"
  >
    <div
      className="col-1 ticker"
    >
      <styled.p>
        Last Price
      </styled.p>
      <styled.p>
        6.0
      </styled.p>
    </div>
    <div
      className="col-1 ticker"
    >
      <styled.p>
        24hr Change
      </styled.p>
      <styled.p
        color="lime"
      >
        4.00
         (
        500.00
        %)
      </styled.p>
    </div>
    <div
      className="col-1 ticker"
    >
      <styled.p>
        24hr High
      </styled.p>
      <styled.p>
        8.0
      </styled.p>
    </div>
    <div
      className="col-1 ticker"
    >
      <styled.p>
        24hr Low
      </styled.p>
      <styled.p>
        9.0
      </styled.p>
    </div>
    <div
      className="col-1 ticker"
    >
      <styled.p>
        24hr Volume
      </styled.p>
      <styled.p>
        7.00
      </styled.p>
    </div>
  </div>
</div>
`;
